ksqldb-functional-tests/src/main/java/io/confluent/ksql/test/driver/AssertExecutor.java

```diff
@@ -47,6 +54,44 @@
  */
 public final class AssertExecutor {
 
+
+  private static final List<SourceProperty> MUST_MATCH = ImmutableList.<SourceProperty>builder()
+      .add(
+          new SourceProperty(
+              DataSource::getSchema,
+              cs -> cs.getElements().toLogicalSchema(),
+              "schema"
+      )).add(
+          new SourceProperty(
+              DataSource::getDataSourceType,
+              cs -> cs instanceof CreateTable ? DataSourceType.KTABLE : DataSourceType.KSTREAM,
+              "type"
+      )).add(
+          new SourceProperty(
+              DataSource::getKafkaTopicName,
+              cs -> cs.getProperties().getKafkaTopic(),
+              "kafka topic"
+      )).add(
+          new SourceProperty(
+              ds -> ds.getKsqlTopic().getValueFormat().getFormatInfo().getFormat(),
+              cs -> cs.getProperties().getValueFormat().name(),
+              "value format"
+          )).add(new SourceProperty(
+              DataSource::getSerdeOptions,
+              cs -> cs.getProperties().getSerdeOptions(),
+              "serde options"
+      )).add(
+          new SourceProperty(
+              ds -> ds.getTimestampColumn().map(TimestampColumn::getColumn),
+              cs -> cs.getProperties().getTimestampColumnName(),
+              "timestamp column"
+      )).add(
+          new SourceProperty(
+              ds -> ds.getTimestampColumn().flatMap(TimestampColumn::getFormat),
+              cs -> cs.getProperties().getTimestampFormat(),
+              "timestamp format"
+      )).build();
```

Comment:
As I'm sure you're aware, this is a very fragile pattern. It's highly likely that people will forget to add things to this list when adding new fields to `DataSource`.

As a good example, `getSerdeOptions` was added recently and must match, but isn't in this list...

Ideally, what we want is for this to 'just work' when new fields are added, or to fail to compile when new fields / properties are added.  OF course, this is tricky without duplicating code!

Not sure of the right approach here.  One partial solution would be to have the code fail if it encounters a `WITH` clause property for which it doesn't have a rule in the above list.  You can probably achieve the same, but better, with a meta test in `AssertExecutorTest` that ensures there is a rule for each property defined in the `CreateConfigs` and `CreateAsConfigs` classes.

Reply:
Makes sense, I'll think about it but it is quite tough... (I don't, for example, have access to the original WITH clause at this point so I don't even know what was typed and what is just resolved to default).

> As a good example, getSerdeOptions was added recently and must match, but isn't in this list...

Let the record show, though, that I did indeed include this in the original code :P (it even shows in the part you highlighted!)

Reply:
> Let the record show, though, that I did indeed include this in the original code :P (it even shows in the part you highlighted!)

Oh year ;)

> I don't, for example, have access to the original WITH clause at this point so I don't even know what was typed and what is just resolved to default

Well, we do have them.  Take `CreateSourceProperties`, it has a `props` field of type `PropertiesConfig`. You can get the exact values in the WITH clause from `props.copyOfOriginalLiterals()` or `props.originals()`.  We'd just need to expose this from `CreateSourceProperties`.

However, I don't think we necessarily need to do that.  A good first step would just be to ensure we have a rule in the above map for each known WITH property.  The list of those can be got from the properties defined in `CreateConfigs.CONFIG_DEF` and `CreateAsConfigs.CONFIG_DEF`. So maybe adding a meta test in `AssertExecutorTest` that compiled the list and then ensured rules exist for all these properties is enough, for now, to give at least _some_ protection from people forgetting to upgrade this when they make changes else where.

Another approach would be to annotate the getters of `DataSource` with something that _execluded_ them from the comparison and then using reflection to do the comparison.  I think this is a better approach.  Happy for this to be done in follow up PR to unblock things, as long as ited tracked and done soon. Otherwise, it's only a matter of time before this diverges from the code its testing.

Reply:
I took your suggestion and found a way to make a test out of it - added the meta-test, thanks for the input!